amethyst_ui/src/text.rs

```diff
@@ -139,503 +134,132 @@ impl Component for TextEditing {
-        Write<'a, UiFocused>,
+        ReadStorage<'a, Selected>,
         Read<'a, EventChannel<Event>>,
-        Read<'a, Time>,
         ReadExpect<'a, ScreenDimensions>,
+        Read<'a, Time>,
     );
 
     fn run(
         &mut self,
-        (entities, mut text, mut editable, transform, mut focused, events, time, screen_dimensions): Self::SystemData,
+        (mut texts, mut text_editings, selecteds, events, screen_dimensions, time): Self::SystemData,
     ) {
-        // Populate and update the tab order cache.
-        {
-            let bitset = &mut self.tab_order_cache.cached;
-            self.tab_order_cache.cache.retain(|&(_t, entity)| {
-                let keep = transform.contains(entity);
-                if !keep {
-                    bitset.remove(entity.id());
-                }
-                keep
-            });
-        }
-
-        for &mut (ref mut t, entity) in &mut self.tab_order_cache.cache {
-            *t = transform
-                .get(entity)
-                .expect("Unreachable: Entities are collected from a prepopulated cache")
-                .tab_order;
-        }
-
-        // Attempt to insert the new entities in sorted position.  Should reduce work during
-        // the sorting step.
-        let transform_set = transform.mask().clone();
-        {
-            // Create a bitset containing only the new indices.
-            let new = (&transform_set ^ &self.tab_order_cache.cached) & &transform_set;
-            for (entity, transform, _new) in (&*entities, &transform, &new).join() {
-                let pos = self
-                    .tab_order_cache
-                    .cache
-                    .iter()
-                    .position(|&(cached_t, _)| transform.tab_order < cached_t);
-                match pos {
-                    Some(pos) => self
-                        .tab_order_cache
-                        .cache
-                        .insert(pos, (transform.tab_order, entity)),
-                    None => self
-                        .tab_order_cache
-                        .cache
-                        .push((transform.tab_order, entity)),
-                }
-            }
-        }
-        self.tab_order_cache.cached = transform_set;
-
-        // Sort from smallest tab order to largest tab order, then by entity creation time.
-        // Most of the time this shouldn't do anything but you still need it for if the tab orders
-        // change.
-        self.tab_order_cache
-            .cache
-            .sort_unstable_by(|&(t1, ref e1), &(t2, ref e2)| {
-                let ret = t1.cmp(&t2);
-                if ret == Ordering::Equal {
-                    return e1.cmp(e2);
-                }
-                ret
-            });
-
-        for text in (&mut text).join() {
+        // Normalize text to ensure we can properly count the characters.
+        // TODO: Possible improvement to be made if this can be moved only when inserting characters into ui text.
```

Comment:
It's also possible the UI text might be programmatically modified by the game developer, so this isn't possible.  I'd already had that thought the first time around 😆

Reply:
I meant having a set_text method ensuring the normalization.

Reply:
We can shelve this for now though, the enhancement is desirable it just needs to be executed carefully.